### PR TITLE
[CodeCompletion] Fast completion for top-level code in single file script

### PR DESCRIPTION
**Pull request by rintaro**
e.g. Playground.
A single file script is like a single function body; the interface of the file does not affect any other files. So when a completion happens in a single file script, re-parse the whole file. But we are still be able to reuse imported modules.

rdar://problem/58378157

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please test

File: lib/IDE/CompletionInstance.cpp [hunk 1]
```diff
@@ -155,10 +155,7 @@ bool CompletionInstance::performCachedOperaitonIfPossible(
     return false;
 
   auto &oldInfo = oldState.getCodeCompletionDelayedDeclState();
-
-  // Currently, only completions within a function body is supported.
-  if (oldInfo.Kind != CodeCompletionDelayedDeclKind::FunctionBody)
```
Comment (benlangmuir) [inline lib/IDE/CompletionInstance.cpp]: Is it worth doing this check later inside the switch when `newInfo.Kind == FunctionBody`?  Or is it covered by the check of `getEquivalentDeclContextFromSourceFile`

File: include/swift/AST/SourceFile.h [hunk 2]
```diff
@@ -344,6 +344,11 @@ class SourceFile final : public FileUnit {
     return BufferID;
   }
 
+  void setBufferID(int Val) {
```
Comment (benlangmuir) [inline include/swift/AST/SourceFile.h]: Why does the `SourceFile` buffer ID need to be updated?  What uses this info during completion?

File: lib/IDE/CompletionInstance.cpp [hunk 3]
```diff
@@ -176,85 +173,124 @@ bool CompletionInstance::performCachedOperaitonIfPossible(
...
+
+    auto *M = oldInfo.ParentContext->getParentModule();
+    if (M->getFiles().size() != 1 || oldSF->Kind != SourceFileKind::Main)
+      return false;
+
+    // Re-parse the whole file. We still re-use imported modules.
+
+    auto sourceText = completionBuffer->getBuffer();
+    if (newInfo.Kind == CodeCompletionDelayedDeclKind::TopLevelCodeDecl) {
+      // We don't need the source text after the top-level code.
```
Comment (benlangmuir) [inline lib/IDE/CompletionInstance.cpp]: How does this work for a whole SourceFile?  Won't slicing the content cause the interface hash to change if you move the cursor past a decl?

File: test/SourceKit/CodeComplete/complete_sequence_toplevel.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  var a: Int
+  var b: Int
+}
+extension Bar {
+  func barMethod() {}
+}
+func foo(arg: Foo) {
+  _ = arg.
+}
+_ = Bar(a: 12, b: 42)
```
Comment (benlangmuir) [inline test/SourceKit/CodeComplete/complete_sequence_toplevel.swift]: We should have some tests that include mutation of top-level expressions (reuse) and top-level decls (no reuse).

File: lib/IDE/CompletionInstance.cpp (hunk 1 above)
Comment (rintaro) [reply lib/IDE/CompletionInstance.cpp]: I believe it covered by interface hash check. If the previous completion happened in toplevel or type members, it must changes the interface hash.

File: include/swift/AST/SourceFile.h (hunk 2 above)
Comment (rintaro) [reply include/swift/AST/SourceFile.h]: At least `ParseMembersRequest::evaluate()` uses `getBufferID()` for delayed member list parsing. https://github.com/apple/swift/blob/53c008f32cf7e83022c9f6643c960015b7f91da7/lib/Parse/ParseRequests.cpp#L36-L44

We can modify this to get bufferID from the decl location, like:
```cpp
  auto &SM = idc->getASTContext().SourceMgr;
  unsigned bufferID = SM.findBufferContainingLoc(idc->getDecl()->getLoc());
```
But I'm not sure this is the only usage of `getBufferID()` during completion.
What is your concern about setting the buffer id?

File: lib/IDE/CompletionInstance.cpp (hunk 3 above)
Comment (rintaro) [reply lib/IDE/CompletionInstance.cpp]: If the completion happens in non-function body position, we don't take interface hash into account. We just re-parse the whole file. So I believe it shouldn't be a problem.

File: test/SourceKit/CodeComplete/complete_sequence_toplevel.swift (hunk 4 above)
Comment (rintaro) [reply test/SourceKit/CodeComplete/complete_sequence_toplevel.swift]: In this patch, for single file script, fast-completion *always* happens as long as compiler argument is not changed. It doesn't matter if it's in expressions or decls.

I will add mutation test cases anyway.

File: include/swift/AST/SourceFile.h (hunk 2 above)
Comment (benlangmuir) [reply include/swift/AST/SourceFile.h]: I guess that's fine, it's just more state that we're mutating so I wanted to understand what effect it has.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/CompletionInstance.cpp [hunk 5]
```diff
@@ -202,85 +199,124 @@ bool CompletionInstance::performCachedOperaitonIfPossible(
...
+      sourceText = sourceText.slice(0, endOffset);
+    }
+    newBufferID = SM.addMemBufferCopy(sourceText,
+                                      completionBuffer->getBufferIdentifier());
+    SM.setCodeCompletionPoint(newBufferID, Offset);
+
+    (void)oldState.takeCodeCompletionDelayedDeclState();
+    oldSF->truncateTopLevelDecls(0);
+    oldSF->clearLookupCache();
+    oldSF->setBufferID(newBufferID);
```
Comment (rintaro) [inline lib/IDE/CompletionInstance.cpp]: It turned out that this doesn't clear the imported modules from the `SourceFile`.
We need to clear them, except for implicitly imported modules.
Or, create a new `SourceFile`, add implicitly imported modules, replace old `SourcFile` with it.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/CompletionInstance.cpp (hunk 5 above)
Comment (rintaro) [reply lib/IDE/CompletionInstance.cpp]: Updated to create a new module and source file.

**Comment by rintaro** [comment]
@benlangmuir Could you take another look?

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test macOS

**Comment by rintaro** [comment]
@swift-ci Please smoke test macOS

**Comment by rintaro** [comment]
@swift-ci Please smoke test OS X platform

**Review by benlangmuir — APPROVED** [review]

**Pull request merged**